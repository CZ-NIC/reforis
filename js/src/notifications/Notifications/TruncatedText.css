.toggle-collapse {
    padding: 0;
}
