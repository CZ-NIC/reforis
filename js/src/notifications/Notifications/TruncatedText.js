import React, { useState } from "react";
import PropTypes from "prop-types";

TruncatedText.propTypes = {
    text: PropTypes.string.isRequired,
    charLimit: PropTypes.number.isRequired,
    threshold: PropTypes.number.isRequired,
};

TruncatedText.defaultProps = {
    threshold: 16,
};

export default function TruncatedText({ text, charLimit, threshold }) {
    const [collapsed, setCollapsed] = useState(true);
    let cardText = text;
    let toggleCollapse = null;

    if (text.length > charLimit + threshold) {
        if (collapsed) {
            cardText = text.substring(0, charLimit).replace(/(\s|\W)$/, "");
            cardText = `${cardText}...`;
        }
        toggleCollapse = <ToggleCollapse collapseFn={setCollapsed} shouldCollapse={!collapsed} />;
    }

    return (
        <>
            <p className="card-text">{cardText}</p>
            {toggleCollapse}
        </>
    );
}

ToggleCollapse.propTypes = {
    collapseFn: PropTypes.func.isRequired,
    shouldCollapse: PropTypes.bool.isRequired,
};

function ToggleCollapse({ collapseFn, shouldCollapse }) {
    return (
        <button
            type="button"
            className="btn btn-link toggle-collapse"
            onClick={() => collapseFn(shouldCollapse)}
        >
            {shouldCollapse ? "Show less" : "Show more"}
        </button>
    );
}
