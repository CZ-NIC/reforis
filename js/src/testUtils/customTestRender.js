/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
