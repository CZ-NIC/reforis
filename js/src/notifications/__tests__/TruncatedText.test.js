/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {render} from 'customTestRender';

import TruncatedText from '../Notifications/TruncatedText';

describe("<TruncatedText/>", () => {
    const ipsum = "Veggies es bonus vobis, proinde vos postulo essum magis kohlrabi welsh onion daikon amaranth tatsoi tomatillo melon azuki bean garlic.";

    it("displays text without show more/less button", () => {
        const { container } = render(<TruncatedText text={ipsum} charLimit={200} />);
        expect(container).toMatchSnapshot();
    });

    it("displays text with show more/less button", () => {
        const { container } = render(<TruncatedText text={ipsum} charLimit={50} />);
        expect(container).toMatchSnapshot();
    });

    it("doesn't truncate the text then limit is exceeded below threshold", () => {
        const { container } = render(<TruncatedText text={ipsum} charLimit={130} />);
        expect(container).toMatchSnapshot();
    });
});
