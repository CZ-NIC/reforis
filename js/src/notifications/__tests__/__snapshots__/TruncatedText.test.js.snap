// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TruncatedText/> displays text with show more/less button 1`] = `
<div>
  <p
    class="card-text"
  >
    Veggies es bonus vobis, proinde vos postulo essum...
  </p>
  <button
    class="btn btn-link toggle-collapse"
    type="button"
  >
    Show more
  </button>
</div>
`;

exports[`<TruncatedText/> displays text without show more/less button 1`] = `
<div>
  <p
    class="card-text"
  >
    Veggies es bonus vobis, proinde vos postulo essum magis kohlrabi welsh onion daikon amaranth tatsoi tomatillo melon azuki bean garlic.
  </p>
</div>
`;

exports[`<TruncatedText/> doesn't truncate the text then limit is exceeded below threshold 1`] = `
<div>
  <p
    class="card-text"
  >
    Veggies es bonus vobis, proinde vos postulo essum magis kohlrabi welsh onion daikon amaranth tatsoi tomatillo melon azuki bean garlic.
  </p>
</div>
`;
